---
title: "Quality control CNS Schirmer"
author: "Stijn Swinnen and Jarne BeliÃ«n"
date: "18-04-2023"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
---

*Laboratory for Neuroimmunology; University of Leuven*


**Based on (for the most part):**

- Satija Lab, [Seurat](https://satijalab.org/seurat): Vignettes.

- University of Cambridge, Wellcome Sanger Institute: [single-cell course](https://www.singlecellcourse.org/index.html).

- Harvard Chan Bioinformatics Core (HBC): [single-cell course](https://hbctraining.github.io/scRNA-seq).

- GitHub: SoupX; Detailed vignette.

- GitHub: scDblFinder; Detailed vignette.

- GitHub: harmony; Detailed vignette.

- Bioconductor: SingleR; Detailed vignette and book.


*References to specific sources (e.g., articles and GitHub community pages) are provided in this script.*


**Overview quality control (QC) parameters:**

- Cell-free (m)RNA contamination removed (SoupX).

- Doublets removed (scDblFinder).

- Number cells in which feature (gene) detected = 5.

- Minimum features (genes) per cell = 200.

- Maximum percentage mitochondrial RNA (mtRNA) per cell = 15 (scRNA-seq) or 5 (snRNA-seq).

- IGHV, IGKV, IGLV, TRAV, TRBV, TRDV and TRGV removed.

- SCTransform used to reduce technical noise.

- Number principal components (PCs) = 30.

- Percentage mtRNA and difference S-G2/M regressed out.

- Batch effect correction and integration with Harmony.

- Resolution clusters = 0.8.


**Required packages (in alphabetical order):**

```{r setup, Print Options}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```


```{r setup, Libraries}

library(celldex)
library(clustree)
library(cowplot)
library(data.table)
library(plyr) # Needs to be loaded before dplyr
library(dplyr)
library(foreach)
library(ggplot2)
library(harmony)
library(pdfCluster)
library(purrr)
library(RColorBrewer)
library(scDblFinder)
library(Seurat)
library(SingleCellExperiment)
library(SingleR)
library(SoupX)
library(stringr)
library(tibble)
library(tidyr)

setwd("L:/GBW-0068_Neuroimmunology/ScRNA-seq/Paper CHIT1/1. Quality control/CNS/Schirmer et al. (2019)")
```


Droplet-based single-cell or single-nucleus RNA-sequencing (scRNA-seq or snRNA-seq) assumes that each **droplet** contains the transcriptome of a singular cell or nucleus; issues to resolve (# 1 and # 2):

- Cell-free (ambient) (m)RNA contamination due to cell lysis.

- Empty droplets; no cell present.

- Doublets/multiplets; multiple cells present.


# 1. Removal cell-free (ambient) (m)RNA contamination due to cell lysis

*Reference SoupX: Young et al. (2020). SoupX removes ambient RNA contamination from droplet-based single-cell RNA-sequencing data. GigaScience.*

- *load10X*. Loads **unfiltered** data (raw count matrices); to profile composition background contamination in empty droplets. Identifies which droplets contain cells (based on Cell Ranger). Removes empty droplets afterwards.

- *autoEstCont*. Calculates contamination fraction at **cluster level**. Default contaminationRange = c(0.01, 0.8). Default priorRho = 0.05 and default priorRhoStdDev (priorRho standard deviation) = 0.1; -> postRho = contamination fraction (most common 0-10%).

- *adjustCounts*. Does not remove genes; subtracts background **counts** for each gene.

```{r, fig.dim = c(10, 7)}

for (i in c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12")) {
  Unfiltered_Data = load10X(paste0("L:/GBW-0068_Neuroimmunology/ScRNA-seq/Paper CHIT1/Data CNS/Schirmer et al. (2019) - 10x Genomics/CNS_1_", i, "/outs")) %>%
    autoEstCont() %>%
    adjustCounts()
  assign(paste0("CNS_1_", i), Unfiltered_Data)
}
```

Output = sparse and **corrected count matrix** to be used instead of original matrix in downstream analysis. Cell count output SoupX = Cell Ranger (filtered count matrices).


# 2. Identification doublets

*Reference scDblFinder: Germain et al. (2021). Doublet identification in single-cell sequencing data using scDblFinder. F1000Research.* 

-> Benchmarked in: *Xi et al. (2021). Benchmarking computational doublet-detection methods for single-cell RNA-sequencing data (in addendum). Cell systems.*

- *scDblFinder*. In particular **heterotypic** doublets. Default clusters = NULL (random artificial doublets = homo- and heterotypic); changed to TRUE (intercluster artificial doublets = heterotypic), because homotypic doublets are unidentifiable based on their transcriptome (do not form intermediate clusters/phenotypes) and therefore erroneous call of actual singlets as doublets is possible. Default artificialDoublets = 5,000. Default dbr (doublet rate) = 1%/1,000 cells (most common 10-20%). Default dbr.sd (doublet rate standard deviation) = 40% dbr.

```{r, fig.dim = c(10, 7)}

CNS_1 <- list(CNS_1_1, CNS_1_2, CNS_1_3, CNS_1_4, CNS_1_5, CNS_1_6, CNS_1_7, CNS_1_8, CNS_1_9, CNS_1_10, CNS_1_11, CNS_1_12)
CNS_1_Names <- list("CNS_1_1", "CNS_1_2", "CNS_1_3", "CNS_1_4", "CNS_1_5", "CNS_1_6", "CNS_1_7", "CNS_1_8", "CNS_1_9", "CNS_1_10", "CNS_1_11", "CNS_1_12")

invisible(foreach(i = CNS_1, j = CNS_1_Names) %do% {
  Doublets <- scDblFinder(i, clusters = TRUE)
  assign(paste0("Doublets_", j), Doublets)
})

Doublets_CNS_1 <- list(Doublets_CNS_1_1, Doublets_CNS_1_2, Doublets_CNS_1_3, Doublets_CNS_1_4, Doublets_CNS_1_5, Doublets_CNS_1_6, Doublets_CNS_1_7, Doublets_CNS_1_8, Doublets_CNS_1_9, Doublets_CNS_1_10, Doublets_CNS_1_11, Doublets_CNS_1_12)

invisible(foreach(i = Doublets_CNS_1, j = CNS_1_Names) %do% {
  Doublets <- data.frame("Barcode" = rownames(colData(i)), "droplet.class" = i$scDblFinder.class, "doublet.score" = i$scDblFinder.score) %>%
    remove_rownames() %>%
    column_to_rownames("Barcode")
  assign(paste0("Doublets_", j), Doublets)
})
head(Doublets)
```

Output = **data frame** to add to metadata Seurat object in # 3.


# 3. Creation Seurat object

- *CreateSeuratObject*. Default assay = "RNA". Default min.cells = 0. Default min.features = 0. It is recommended to first use low values for min.cells and min.features; retain most features (genes) and cells (https://github.com/satijalab/seurat/issues/3812). In-depth quality control (QC) will ensue in # 4.

```{r, fig.dim = c(10, 7)}

invisible(foreach(i = CNS_1, j = CNS_1_Names) %do% {
  Seurat_Object <- CreateSeuratObject(counts = i, project = j, min.cells = 5, min.features = 100)
  assign(j, Seurat_Object)
})
```

Allocate droplet class ("singlet" or "doublet") and doublet score:
```{r, fig.dim = c(10, 7)}

CNS_1_1 <- AddMetaData(CNS_1_1, Doublets_CNS_1_1)
CNS_1_2 <- AddMetaData(CNS_1_2, Doublets_CNS_1_2)
CNS_1_3 <- AddMetaData(CNS_1_3, Doublets_CNS_1_3)
CNS_1_4 <- AddMetaData(CNS_1_4, Doublets_CNS_1_4)
CNS_1_5 <- AddMetaData(CNS_1_5, Doublets_CNS_1_5)
CNS_1_6 <- AddMetaData(CNS_1_6, Doublets_CNS_1_6)
CNS_1_7 <- AddMetaData(CNS_1_7, Doublets_CNS_1_7)
CNS_1_8 <- AddMetaData(CNS_1_8, Doublets_CNS_1_8)
CNS_1_9 <- AddMetaData(CNS_1_9, Doublets_CNS_1_9)
CNS_1_10 <- AddMetaData(CNS_1_10, Doublets_CNS_1_10)
CNS_1_11 <- AddMetaData(CNS_1_11, Doublets_CNS_1_11)
CNS_1_12 <- AddMetaData(CNS_1_12, Doublets_CNS_1_12)
```

Allocate sequencing technique ("scRNA-seq" or "snRNA-seq"):
```{r, fig.dim = c(10, 7)}

CNS_1_1[["sequencing.technique"]] <- "snRNA-seq"
CNS_1_2[["sequencing.technique"]] <- "snRNA-seq"
CNS_1_3[["sequencing.technique"]] <- "snRNA-seq"
CNS_1_4[["sequencing.technique"]] <- "snRNA-seq"
CNS_1_5[["sequencing.technique"]] <- "snRNA-seq"
CNS_1_6[["sequencing.technique"]] <- "snRNA-seq"
CNS_1_7[["sequencing.technique"]] <- "snRNA-seq"
CNS_1_8[["sequencing.technique"]] <- "snRNA-seq"
CNS_1_9[["sequencing.technique"]] <- "snRNA-seq"
CNS_1_10[["sequencing.technique"]] <- "snRNA-seq"
CNS_1_11[["sequencing.technique"]] <- "snRNA-seq"
CNS_1_12[["sequencing.technique"]] <- "snRNA-seq"
```

Allocate subject ("MS_0...", "MS_1...", "MS_2...", "MS_3..." or "MS_4..."):
```{r, fig.dim = c(10, 7)}

CNS_1_1[["subject"]] <- "MS_1_1"
CNS_1_2[["subject"]] <- "MS_1_2"
CNS_1_3[["subject"]] <- "MS_1_3"
CNS_1_4[["subject"]] <- "MS_1_4"
CNS_1_5[["subject"]] <- "MS_1_5"
CNS_1_6[["subject"]] <- "MS_1_6"
CNS_1_7[["subject"]] <- "MS_1_7"
CNS_1_8[["subject"]] <- "MS_1_8"
CNS_1_9[["subject"]] <- "MS_1_9"
CNS_1_10[["subject"]] <- "MS_1_9"
CNS_1_11[["subject"]] <- "MS_1_4"
CNS_1_12[["subject"]] <- "MS_1_10"
```

Allocate disease course ("RRMS", "PPMS", "SPMS", "PMS", "MS" or "NA"):
```{r, fig.dim = c(10, 7)}

CNS_1_1[["disease.course"]] <- "SPMS"
CNS_1_2[["disease.course"]] <- "SPMS"
CNS_1_3[["disease.course"]] <- "SPMS"
CNS_1_4[["disease.course"]] <- "SPMS"
CNS_1_5[["disease.course"]] <- "SPMS"
CNS_1_6[["disease.course"]] <- "SPMS"
CNS_1_7[["disease.course"]] <- "PPMS"
CNS_1_8[["disease.course"]] <- "SPMS"
CNS_1_9[["disease.course"]] <- "SPMS"
CNS_1_10[["disease.course"]] <- "SPMS"
CNS_1_11[["disease.course"]] <- "SPMS"
CNS_1_12[["disease.course"]] <- "SPMS"
```

Allocate tissue ("CSF", "Brain", "Spinal cord" or "NA"):
```{r, fig.dim = c(10, 7)}

CNS_1_1[["tissue"]] <- "Brain"
CNS_1_2[["tissue"]] <- "Brain"
CNS_1_3[["tissue"]] <- "Brain"
CNS_1_4[["tissue"]] <- "Brain"
CNS_1_5[["tissue"]] <- "Brain"
CNS_1_6[["tissue"]] <- "Brain"
CNS_1_7[["tissue"]] <- "Brain"
CNS_1_8[["tissue"]] <- "Brain"
CNS_1_9[["tissue"]] <- "Brain"
CNS_1_10[["tissue"]] <- "Brain"
CNS_1_11[["tissue"]] <- "Brain"
CNS_1_12[["tissue"]] <- "Brain"
```

Allocate tissue subtype ("White matter", "Grey matter", "White and grey matter" or "NA"):
```{r, fig.dim = c(10, 7)}

CNS_1_1[["tissue.subtype"]] <- "White and grey matter"
CNS_1_2[["tissue.subtype"]] <- "White and grey matter"
CNS_1_3[["tissue.subtype"]] <- "White and grey matter"
CNS_1_4[["tissue.subtype"]] <- "White and grey matter"
CNS_1_5[["tissue.subtype"]] <- "White and grey matter"
CNS_1_6[["tissue.subtype"]] <- "White and grey matter"
CNS_1_7[["tissue.subtype"]] <- "White and grey matter"
CNS_1_8[["tissue.subtype"]] <- "White and grey matter"
CNS_1_9[["tissue.subtype"]] <- "White and grey matter"
CNS_1_10[["tissue.subtype"]] <- "White and grey matter"
CNS_1_11[["tissue.subtype"]] <- "White and grey matter"
CNS_1_12[["tissue.subtype"]] <- "White and grey matter"
```

Allocate pathology ("Lesion", "Lesion core", "Lesion edge", "Periplaque", "Normal-appearing matter" or "NA"):
```{r, fig.dim = c(10, 7)}

CNS_1_1[["pathology"]] <- "Lesion"
CNS_1_2[["pathology"]] <- "Lesion"
CNS_1_3[["pathology"]] <- "Lesion"
CNS_1_4[["pathology"]] <- "Lesion"
CNS_1_5[["pathology"]] <- "Lesion"
CNS_1_6[["pathology"]] <- "Lesion"
CNS_1_7[["pathology"]] <- "Lesion"
CNS_1_8[["pathology"]] <- "Lesion"
CNS_1_9[["pathology"]] <- "Lesion"
CNS_1_10[["pathology"]] <- "Lesion"
CNS_1_11[["pathology"]] <- "Lesion"
CNS_1_12[["pathology"]] <- "Lesion"
```

Allocate pathology activity ("Acute active", "Acute inactive", "Chronic active", "Chronic inactive", "Acute/chronic active", "Acute/chronic inactive" or "NA"):
```{r, fig.dim = c(10, 7)}

CNS_1_1[["pathology.activity"]] <- "Acute/chronic active"
CNS_1_2[["pathology.activity"]] <- "Chronic inactive"
CNS_1_3[["pathology.activity"]] <- "Chronic inactive"
CNS_1_4[["pathology.activity"]] <- "Acute/chronic active"
CNS_1_5[["pathology.activity"]] <- "Chronic inactive"
CNS_1_6[["pathology.activity"]] <- "Acute/chronic active"
CNS_1_7[["pathology.activity"]] <- "Acute/chronic active"
CNS_1_8[["pathology.activity"]] <- "Acute/chronic active"
CNS_1_9[["pathology.activity"]] <- "Acute/chronic active"
CNS_1_10[["pathology.activity"]] <- "Acute/chronic active"
CNS_1_11[["pathology.activity"]] <- "Acute/chronic active"
CNS_1_12[["pathology.activity"]] <- "Chronic inactive"

CNS_1 <- list(CNS_1_1, CNS_1_2, CNS_1_3, CNS_1_4, CNS_1_5, CNS_1_6, CNS_1_7, CNS_1_8, CNS_1_9, CNS_1_10, CNS_1_11, CNS_1_12)

for (i in CNS_1) {
  summary(i@meta.data) %>%
    print()
}
```

Output = **Seurat objects** with assay **RNA** and slot **counts** (raw). Next, we **merge** these separate Seurat objects into a singular Seurat object. Because the same bar code can be used in different samples, we add a sample-specific prefix to each bar code.

```{r, fig.dim = c(10, 7)}

Merged_CNS_1 <- merge(CNS_1_1, c(CNS_1_2, CNS_1_3, CNS_1_4, CNS_1_5, CNS_1_6, CNS_1_7, CNS_1_8, CNS_1_9, CNS_1_10, CNS_1_11, CNS_1_12), add.cell.ids = CNS_1_Names)

head(Merged_CNS_1)
tail(Merged_CNS_1)
summary(Merged_CNS_1@meta.data)
DefaultAssay(Merged_CNS_1)
```


# 4. Determination quality control (QC) parameters to identify high-quality cells


Let us now define **High-quality** cells:

- High number features (genes)/cell.

- Low percentage mitochondrial RNA (mtRNA)/cell.

- Droplet class = singlets.

First, let's calculate the percentage of mitochondrial RNA (mtRNA) and ribosomal (rRNA):
```{r, fig.dim = c(10, 7)}

Merged_CNS_1[["percent.mt"]] <- PercentageFeatureSet(Merged_CNS_1, pattern = "^MT-")
Merged_CNS_1[["percent.rb"]] <- PercentageFeatureSet(Merged_CNS_1, pattern = "^RP[SL]")

summary(Merged_CNS_1@meta.data)
```

Next we plot these features:

FeatureScatter:
```{r, fig.dim = c(10, 7)}

for (i in CNS_1_Names) {
  Subset <- subset(Merged_CNS_1, orig.ident == i)
  print(i)
  summary(Subset@meta.data) %>%
    print()
}

VlnPlot(Merged_CNS_1, features = "nCount_RNA")
VlnPlot(Merged_CNS_1, features = "nFeature_RNA")
VlnPlot(Merged_CNS_1, features = "percent.mt")
VlnPlot(Merged_CNS_1, features = "percent.rb")
VlnPlot(Merged_CNS_1, features = "doublet.score")
```

Correlation QC parameters:
```{r, fig.dim = c(10, 7)}

Plot1 <- FeatureScatter(Merged_CNS_1, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
Plot2 <- FeatureScatter(Merged_CNS_1, feature1 = "nFeature_RNA", feature2 = "percent.mt")
Plot3 <- FeatureScatter(Merged_CNS_1, feature1 = "nFeature_RNA", feature2 = "percent.rb")
Plot4 <- FeatureScatter(Merged_CNS_1, feature1 = "percent.mt", feature2 = "percent.rb")
Plot5 <- FeatureScatter(Merged_CNS_1, feature1 = "nFeature_RNA", feature2 = "doublet.score")

Plot1
Plot2
Plot3
Plot4
Plot5
```

Now, we add a QC column (as a subset; no low-quality cells removed yet) to the metadata of the merged Seurat object and specify our **pass or fail criteria**. We will implement these QC criteria in # 8.

*Reference percentage mtRNA: Absinta et al. (2021). A lymphocyte-microglia-astrocyte axis in chronic active multiple sclerosis. Nature* and *Trobisch et al. (2022). Cross-regional homeostatic and reactive glial signatures in multiple sclerosis. Acta Neuropathologica.* -> **< 5%**.

```{r, fig.dim = c(10, 7)}

Merged_CNS_1[["quality.control"]] <- ifelse(Merged_CNS_1$nFeature_RNA > 200 & Merged_CNS_1$percent.mt < 5 & Merged_CNS_1$droplet.class == "singlet", "Pass", "Fail")
```

Overview "Fail" and "Pass" per sample:
```{r, fig.dim = c(10, 7)}

Table1 <- table(Merged_CNS_1$orig.ident, Merged_CNS_1$quality.control)
Table2 <- colSums(Table1)
Table3 <- prop.table(Table2)
Table4 <- sum(Table3)
Table5 <- prop.table(Table1, margin = 1)
Table6 <- rowSums(Table5)

Table1 # "Fail" and "Pass" per sample
Table2 # Sum total "Fail" and "Pass"
Table3 # Proportion total "Fail" and "Pass"
Table4 # Sum proportions total "Fail" and "Pass"; must be 1
Table5 # Proportion "Fail" and "Pass" per sample
Table6 # Sum proportions "Fail" and "Pass" per sample; must be 1

summary(subset(Merged_CNS_1@meta.data, quality.control == "Pass"))

for (i in CNS_1_Names) {
  Subset <- subset(Merged_CNS_1, quality.control == "Pass" & orig.ident == i)
  print(i)
  summary(Subset@meta.data) %>%
    print()
}

VlnPlot(subset(Merged_CNS_1, quality.control == "Pass"), features = "nFeature_RNA")
VlnPlot(subset(Merged_CNS_1, quality.control == "Pass"), features = "percent.mt")
VlnPlot(subset(Merged_CNS_1, quality.control == "Pass"), features = "percent.rb")
VlnPlot(subset(Merged_CNS_1, quality.control == "Pass"), features = "doublet.score")
```


Sources (unwanted) **variation** (# 5, # 6 and # 7):

- Variable fragments B and T cell receptors (BCR and TCR); induced subject-specific clusters in our hands.

- Differences cell cycle phase.

- Batch effects.


# 5. Removal variable fragments B and T cell receptors (BCR and TCR)

```{r, fig.dim = c(10, 7)}

Counts <- GetAssayData(Merged_CNS_1, assay = "RNA", slot = "counts")
Genes <- rownames(Counts)
Variable_Genes <- Genes[str_detect(Genes, pattern = "^IGHV|^IGKV|^IGLV|^TRAV|^TRBV|^TRDV|^TRGV")]
Counts <- Counts[-(which(rownames(Counts) %in% Variable_Genes)), ]
Merged_CNS_1 <- subset(Merged_CNS_1, features = rownames(Counts))

Variable_Genes <- as.data.frame(Variable_Genes)
dplyr::count(Variable_Genes) # Number genes removed
Genes <- as.data.frame(Genes)
dplyr::count(Variable_Genes)/dplyr::count(Genes) # Proportion genes removed
```

Influence on number cells "Fail" or "Pass" QC criteria?
```{r, fig.dim = c(10, 7)}

Merged_CNS_1[["quality.control"]] <- ifelse(Merged_CNS_1$nFeature_RNA > 200 & Merged_CNS_1$percent.mt < 5 & Merged_CNS_1$droplet.class == "singlet", "Pass", "Fail")
Table1 <- table(Merged_CNS_1$orig.ident, Merged_CNS_1$quality.control)
Table1
```


# 6. Exploration cell cycle phase variation

It is recommended to **normalize** the data before assessment of the cell cycle score (https://github.com/satijalab/seurat/issues/1679). The cell cycle score is assigned to each **individual cell** - independent of other cells -, which means this function can be used before implementation of the QC criteria in # 8 (i.e., no influence of cells that will be removed).

- *NormalizeData*. Corrects for sequencing depth; "Merged_CNS_1@assays$RNA@data". Default normalization.method = "LogNormalize". Default scale.factor = 10,000 (as if all cells contain 10,000 transcripts).

- *FindVariableFeatures*. For principal component analysis (PCA). Default selection.method = "vst". Default nfeatures = 2,000.

- *ScaleData*. Corrects for differences in variation gene expression, Z-score; "Merged_CNS_1@assays$RNA@scale.data". Mean expression = 0; expression variance across cells = 1 (i.e., equal variation gene expression). Default features = NULL (all genes).

```{r, fig.dim = c(10, 7)}

Merged_CNS_1 <- NormalizeData(Merged_CNS_1, verbose = FALSE)
Merged_CNS_1 <- FindVariableFeatures(Merged_CNS_1, verbose = FALSE)
Top10 <- head(VariableFeatures(Merged_CNS_1), 10)
Top10 # Unless removed, dominated by variable fragments BCR and TCR
Plot <- VariableFeaturePlot(Merged_CNS_1)
LabelPoints(plot = Plot, points = Top10, repel = TRUE, xnudge = 0, ynudge = 0)
Merged_CNS_1 <- ScaleData(Merged_CNS_1, features = rownames(Merged_CNS_1), verbose = FALSE)
```

Output = **Seurat object** with assay **RNA** and slots **counts** (raw), **data** (log-normalized) and **scale.data** (scaled).

Next, we evaluate cell cycle scores in our cells:
```{r, fig.dim = c(10, 7)}

cc.genes.updated.2019
s.genes <- cc.genes.updated.2019$s.genes
g2m.genes <- cc.genes.updated.2019$g2m.genes
Merged_CNS_1 <- CellCycleScoring(Merged_CNS_1, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE) # Active.ident changed from samples (orig.ident) to cell cycle phases

summary(Merged_CNS_1@meta.data)
```

Cell cycle statistics:
```{r, fig.dim = c(10, 7)}

Table1 <- table(Merged_CNS_1$orig.ident, Merged_CNS_1$Phase)
Table2 <- prop.table(Table1, margin = 1)
Table3 <- rowSums(Table2)

Table1 # Cell cycle phases per sample
Table2 # Proportion cell cycle phases per sample
Table3 # Sum proportions cell cycle phases per sample; must be 1
```

Since we only use cell cycle genes as input for the principal components (PCs), there is an insufficient number of genes to generate 50 PCs (default); changed to 35 PCs (https://github.com/satijalab/seurat/discussions/4567):
```{r, fig.dim = c(10, 7)}

Merged_CNS_1 <- RunPCA(Merged_CNS_1, features = c(s.genes, g2m.genes), npcs = 35)
Merged_CNS_1@reductions
DefaultDimReduc(Merged_CNS_1)
PCAPlot(Merged_CNS_1) # Displays variation between cells due to differences in cell cycle phase
Merged_CNS_1 <- RunPCA(Merged_CNS_1)
Merged_CNS_1@reductions
DefaultDimReduc(Merged_CNS_1)
PCAPlot(Merged_CNS_1, split.by = "Phase") # Displays influence cell cycle genes on PC_1 and PC_2
ElbowPlot(Merged_CNS_1)
Merged_CNS_1 <- FindNeighbors(Merged_CNS_1)
Merged_CNS_1 <- FindClusters(Merged_CNS_1) # Active.ident changed from cell cycle phases to clusters
Merged_CNS_1 <- RunUMAP(Merged_CNS_1, dims = 1:15, verbose = FALSE)
Merged_CNS_1@reductions
DefaultDimReduc(Merged_CNS_1)
UMAPPlot(Merged_CNS_1, label = TRUE)
FeaturePlot(Merged_CNS_1, features = "S.Score", min.cutoff = "q10", label = TRUE)
VlnPlot(Merged_CNS_1, features = "S.Score")
FeaturePlot(Merged_CNS_1, features = "G2M.Score", min.cutoff = "q10", label = TRUE)
VlnPlot(Merged_CNS_1, features = "G2M.Score")
```

There might be some biologically relevant variation between **cycling** - S and G2M genes - and **non-cycling cells** - G1 genes - (e.g. plasmablasts). Thus, we want to keep this source of variation in our data. The difference between cells with expression of S genes or G2M genes, however, is minimal (and probably an unwanted source of variation). Therefore, we regress this difference out in # 9.

```{r, fig.dim = c(10, 7)}

Merged_CNS_1[["cc.difference"]] <- Merged_CNS_1$S.Score - Merged_CNS_1$G2M.Score
summary(Merged_CNS_1@meta.data)
```


# 7. Estimation batch effects

Visualization of the batch effects:
```{r, fig.dim = c(10, 7)}

# UMAPPlot(Merged_CNS_1, group.by = "batch", label = TRUE, repel = TRUE)
# There is no batch information available for this data set
```


# 8. Implementation QC criteria

Which cells will be removed (some cells might be removed due to more than one criterion) under our QC criteria?

- **nFeature_RNA < 200**
```{r, fig.dim = c(10, 7)}

Fail_nFeature_RNA <- WhichCells(Merged_CNS_1, expression = nFeature_RNA < 200)
dplyr::count(as.data.frame(Fail_nFeature_RNA))
Subset <- subset(Merged_CNS_1, cells = Fail_nFeature_RNA)
Table1 <- table(Subset$old.ident, Subset$quality.control)
Table2 <- Table1/as.vector(table(Merged_CNS_1$old.ident))
Table3 <- table(Subset$RNA_snn_res.0.8, Subset$quality.control)
Table4 <- Table3/as.vector(table(Merged_CNS_1$RNA_snn_res.0.8))

Table1 # "Fail" nFeature_RNA per sample
Table2 # Proportion "Fail" nFeature_RNA per sample
Table3 # "Fail" nFeature_RNA per cluster
Table4 # Proportion "Fail" nFeature_RNA per cluster

UMAPPlot(Merged_CNS_1, split.by = "quality.control", label = TRUE, cells.highlight = Fail_nFeature_RNA) + NoLegend()
```

- **percent.mt > 5**
```{r, fig.dim = c(10, 7)}

Fail_percent.mt <- WhichCells(Merged_CNS_1, expression = percent.mt > 5)
dplyr::count(as.data.frame(Fail_percent.mt))
Subset <- subset(Merged_CNS_1, cells = Fail_percent.mt)
Table1 <- table(Subset$old.ident, Subset$quality.control)
Table2 <- Table1/as.vector(table(Merged_CNS_1$old.ident))
Table3 <- table(Subset$RNA_snn_res.0.8, Subset$quality.control)
Table4 <- Table3/as.vector(table(Merged_CNS_1$RNA_snn_res.0.8))

Table1 # "Fail" percent.mt per sample
Table2 # Proportion "Fail" percent.mt per sample
Table3 # "Fail" percent.mt per cluster
Table4 # Proportion "Fail" percent.mt per cluster

UMAPPlot(Merged_CNS_1, split.by = "quality.control", label = TRUE, cells.highlight = Fail_percent.mt) + NoLegend()
```

- **droplet.class == "doublet"**
```{r, fig.dim = c(10, 7)}

Fail_droplet.class <- WhichCells(Merged_CNS_1, expression = droplet.class == "doublet")
dplyr::count(as.data.frame(Fail_droplet.class))
Subset <- subset(Merged_CNS_1, cells = Fail_droplet.class)
Table1 <- table(Subset$old.ident, Subset$quality.control)
Table2 <- Table1/as.vector(table(Merged_CNS_1$old.ident))
Table3 <- table(Subset$RNA_snn_res.0.8, Subset$quality.control)
Table4 <- Table3/as.vector(table(Merged_CNS_1$RNA_snn_res.0.8))

Table1 # "Fail" droplet.class per sample
Table2 # Proportion "Fail" droplet.class per sample
Table3 # "Fail" droplet.class per cluster
Table4 # Proportion "Fail" droplet.class per cluster

UMAPPlot(Merged_CNS_1, split.by = "quality.control", label = TRUE, cells.highlight = Fail_droplet.class) + NoLegend()
```

Overview "Fail" and "Pass" per cluster (cluster information was not yet available in # 4):
```{r, fig.dim = c(10, 7)}

Table1 <- table(Merged_CNS_1$RNA_snn_res.0.8, Merged_CNS_1$quality.control)
Table2 <- colSums(Table1)
Table3 <- prop.table(Table2)
Table4 <- sum(Table3)
Table5 <- prop.table(Table1, margin = 1)
Table6 <- rowSums(Table5)

Table1 # "Fail" and "Pass" per cluster
Table2 # Sum total "Fail" and "Pass"
Table3 # Proportion total "Fail" and "Pass"
Table4 # Sum proportions total "Fail" and "Pass"; must be 1
Table5 # Proportion "Fail" and "Pass" per cluster
Table6 # Sum proportions "Fail" and "Pass" per cluster; must be 1
```

Implementation QC criteria:
```{r, fig.dim = c(10, 7)}

Merged_CNS_1 <- subset(Merged_CNS_1, quality.control == "Pass")

summary(Merged_CNS_1@meta.data)

for (i in CNS_1_Names) {
  Subset <- subset(Merged_CNS_1, orig.ident == i)
  print(i)
  summary(Subset@meta.data) %>%
    print()
}
```


# 9. Data normalization and variance stabilization with *SCTransform*

*Reference SCTransform: Hafemeister and Satija (2019). Normalization and variance stabilization of single-cell RNA-seq data using regularized negative binomial regression. Genome Biology.*

*SCTransform* is beneficial to find **"rare" (low-frequency) cell populations**. In comparison to *NormalizeData*, *SCTransform* is less susceptible to technical noise and therefore allows more variable features (genes; 2,000 -> 3,000) and more PCs. In addition, we **regress out** (not remove) the percentage of mtRNA (common practice) and the difference between the expression of S and G2M genes (difference between cycling and non-cycling cells remains).

- *SCTransform*. *NormalizeData*, *FindVariableFeatures* and *ScaleData* in one optimized function; "Split_CNS_1@assays$SCT". Default assay = "RNA". Default residual.features = NULL (all genes). Default variable.features.n = 3,000.

We split the merged Seurat object again before *SCTransfrom*: (https://github.com/satijalab/seurat/issues/5306)
```{r, fig.dim = c(10, 7)}
  rm(Seurat_Object) #To create some more space before creating another onject
  rm(CNS_1_1,CNS_1_2,CNS_1_3,CNS_1_4,CNS_1_5,CNS_1_6,CNS_1_7,CNS_1_8,CNS_1_9,CNS_1_10,CNS_1_11,CNS_1_12)
  rm(Subset)
Split_CNS_1 <- SplitObject(Merged_CNS_1, split.by = "orig.ident")
  rm(Merged_CNS_1)
for (i in 1:length(Split_CNS_1)) {
  Split_CNS_1[[i]] <- SCTransform(Split_CNS_1[[i]], vars.to.regress = c("percent.mt", "cc.difference"), verbose = FALSE)
}
```

Check assays stored in Seurat objects (e.g., CNS_1_1):
```{r, fig.dim = c(10, 7)}

Split_CNS_1$CNS_1_1@assays
DefaultAssay(Split_CNS_1$CNS_1_1)
```

Output = **Seurat objects** with additional assay **SCT** and slots **counts** (corrected; reverse-transformed from Pearson residuals in scale.data and sequence-depth corrected), **data** (corrected; log-normalized counts) and **scale.data** (Pearson residuals).


# 10. Data integration and batch effect correction with Harmony

*Reference Harmony: Korsunsky et al (2019). Fast, sensitive and accurate integration of single-cell data with Harmony. Nature Methods.*

We specify that we used **SCTransform** with 3,000 variable features (genes).

- *SelectIntegrationFeatures*. Default nfeatures = 2,000.

- *RunPCA*. Default features = NULL (variable features). Default npcs (number of PCs) = 50.

Harmony integration transforms principal components (PCs) and seeks clusters with maximum "diversity" based on k-means, i.e. whether each cluster contains balanced amounts of cells from each batch (e.g., donor, condition, tissue,...). Harmony tries to correct the imbalance or **batch effects**. To do so, a merged and normalized Seurat object with defined variable features and PCs is required (https://hbctraining.github.io/scRNA-seq_online/lessons/06a_integration_harmony.html).

-> Benchmarked in *Tran et al. (2020). A benchmark of batch-effect correction methods for single-cell RNA-sequencing data. Genome Biology.* and *Luecken et al. (2022). Benchmarking atlas-level data integration in single-cell genomics. Nature Methods.* 

We treat each **sample** as a different batch in order to integrate on the level of the individual samples.
```{r, fig.dim = c(10, 7)}

Integration_Features <- SelectIntegrationFeatures(Split_CNS_1, nfeatures = 3000)
Merged_CNS_1 <- merge(Split_CNS_1[[1]], Split_CNS_1[2:length(Split_CNS_1)])
VariableFeatures(Merged_CNS_1) <- Integration_Features
Merged_CNS_1 <- RunPCA(Merged_CNS_1, assay = "SCT")

Harmonized_CNS_1 <- RunHarmony(Merged_CNS_1, group.by.vars = "orig.ident", reduction = "pca", assay.use = "SCT", reduction.save = "harmony")
Harmonized_CNS_1@reductions
DefaultDimReduc(Harmonized_CNS_1)
DefaultAssay(Harmonized_CNS_1)
```

Output = batch-corrected **Seurat object** with additional reduction **harmony** (in addition to **pca**).


# 11. Clustering

Selection of the number of "Harmony" PCs:

- *ElbowPlot*. Default reduction = "pca".
```{r, fig.dim = c(10, 7)}

ElbowPlot(Harmonized_CNS_1, ndims = 50, reduction = "harmony")
```

*SCTransform* is less susceptible to technical noise than *NormalizeData*, *FindVariableFeatures* and *ScaleData*. Therefore, we can choose a higher number of "Harmony" PCs without incorporating - too much - technical sources of variation. We decide to use the first **30** "Harmony" PCs.

- *FindNeighbors*. Default reduction = "pca".

- *FindClusters*. Default resolution = 0.8.

- *RunUMAP*. Default reduction = "pca".

```{r, fig.dim = c(10, 7)}

Harmonized_CNS_1 <- FindNeighbors(Harmonized_CNS_1, reduction = "harmony", dims = 1:30)
Harmonized_CNS_1 <- FindClusters(Harmonized_CNS_1)
Harmonized_CNS_1@commands$FindClusters # Parameters used to determine clusters
Harmonized_CNS_1 <- RunUMAP(Harmonized_CNS_1, reduction = "harmony", dims = 1:30, verbose = FALSE)
Harmonized_CNS_1@reductions
DefaultDimReduc(Harmonized_CNS_1)
UMAPPlot(Harmonized_CNS_1, label = TRUE) # Resolution 0.8
```

Visualization batch effects post-integration:
```{r, fig.dim = c(10, 7)}

# UMAPPlot(Harmonized_CNS_4, group.by = "batch", label = TRUE, repel = TRUE)
# There is no batch information available for this data set
```


# 12. Preliminary broad cluster annotation

We probe whether we find the **expected (large) cell populations**. With regard to differential gene expression (DGE) and visualization of differentially expressed genes (DEGs), it is recommended to use the **RNA assay** - not the SCT assay (https://github.com/satijalab/seurat/issues/4081; https://github.com/satijalab/seurat/issues/4082).


## 12.1 Manual annotation

```{r, fig.dim = c(10, 7)}

DefaultAssay(Harmonized_CNS_1)
DefaultAssay(Harmonized_CNS_1) <- "RNA"
DefaultAssay(Harmonized_CNS_1)
```

**Myeloid cells**

- General myeloid cells: Upregulation PTPRC, CSF1R, P2RY12, ITGAM and LYZ.

```{r, fig.dim = c(10, 7)}

FeaturePlot(Harmonized_CNS_1, features = c("PTPRC", "CSF1R", "P2RY12", "ITGAM", "LYZ"), label = TRUE)
VlnPlot(Harmonized_CNS_1, features = c("PTPRC", "CSF1R", "P2RY12", "ITGAM", "LYZ"))
```

**Dendritic cells**

- Plasmacytoid dendritic cells (pDC): Upregulation TCF4 and TNFRSF21.

- Myeloid dendritic cells 1 (mDC 1): Upregulation WDFY4, XCR1, BATF3, AXL and SIGLEC6.

- Myeloid dendritic cells 2 (mDC 2): Upregulation FCER1A, CD1C, CLEC10A and CLEC9A.

```{r, fig.dim = c(10, 7)}

FeaturePlot(Harmonized_CNS_1, features = c("TCF4", "TNFRSF21"), label = TRUE)
FeaturePlot(Harmonized_CNS_1, features = c("WDFY4", "XCR1", "BATF3", "AXL", "SIGLEC6"), label = TRUE)
FeaturePlot(Harmonized_CNS_1, features = c("FCER1A", "CD1C", "CLEC10A", "CLEC9A"), label = TRUE)
```

**Granulocytes**

- Granulocytes: Upregulation S100A8 and S100A9.

```{r, fig.dim = c(10, 7)}

FeaturePlot(Harmonized_CNS_1, features = c("S100A8", "S100A9"), label = TRUE)
```

**Monocytes and macrophages/microglia**

- General monocytes: Upregulation VCAN, CCR2, HLA-DBP1 and CD74.

- CD14+ monocytes: Upregulation CD14.

- CD16+ monocytes: Upregulation FCGR3A (CD16).

- Macrophages/microglia: Upregulation APOE and TREM2.

- Macrophages: Upregulation EMP3 and LYVE1.

- Microglia: Upregulation CD81, P2RY12, TMEM119, SPI1 and CX3CR1.

```{r, fig.dim = c(10, 7)}

FeaturePlot(Harmonized_CNS_1, features = c("VCAN", "CCR2", "HLA-DBP1", "CD74"), label = TRUE)
FeaturePlot(Harmonized_CNS_1, features = "CD14", label = TRUE)
FeaturePlot(Harmonized_CNS_1, features = "FCGR3A", label = TRUE)
FeaturePlot(Harmonized_CNS_1, features = c("APOE", "TREM2"), label = TRUE)
VlnPlot(Harmonized_CNS_1, features = c("APOE", "TREM2"))
FeaturePlot(Harmonized_CNS_1, features = c("EMP3", "LYVE1"), label = TRUE)
FeaturePlot(Harmonized_CNS_1, features = c("CD81", "P2RY12", "TMEM119", "SPI1", "CX3CR1"), label = TRUE)
```


## 12.2 Automated cluster annotation with *SingleR*

*Reference SingleR: Aran et al. (2019). Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage. Nature immunology.*

We compare our manual cluster annotation (in # 12.1) with an automated cluster annotation method *SingleR* - based on Spearman correlations between our dataset and a reference dataset. We use a **reference dataset** from *celldex* (Novershtern (microarray)).

Immune reference dataset (celldex):
```{r, fig.dim = c(10, 7)}

Novershtern.ref <- celldex::NovershternHematopoieticData()
```

Convert Seurat object to SingleCellExperiment (Bioconductor):
```{r, fig.dim = c(10, 7)}

SCE <- as.SingleCellExperiment(DietSeurat(Harmonized_CNS_1))
SCE
```

Extract cluster information:
```{r, fig.dim = c(10, 7)}

Clusters <- SCE$seurat_clusters
```

Cluster annotation with fine labels:
```{r, fig.dim = c(10, 7)}

Novershtern.fine <- SingleR(test = SCE, ref = Novershtern.ref, labels = Novershtern.ref$label.fine, clusters = Clusters)
table(Novershtern.fine$labels) # Number clusters with same annotation; grouped
table(Novershtern.fine$pruned.labels) # Low-quality annotations removed
plotScoreHeatmap(Novershtern.fine, show.pruned = TRUE, show_colnames = TRUE)
Harmonized_CNS_1[["Novershtern.fine"]] <- Novershtern.fine$labels[match(Harmonized_CNS_1[[]]["seurat_clusters"]$seurat_clusters, rownames(Novershtern.fine))]
Harmonized_CNS_1 <- SetIdent(Harmonized_CNS_1, value = "Novershtern.fine")
UMAPPlot(Harmonized_CNS_1, label = TRUE) + NoLegend()
```

Cell annotation with fine labels:
```{r, fig.dim = c(10, 7)}
Novershtern.fine <- SingleR(test = SCE, ref = Novershtern.ref, labels = Novershtern.ref$label.fine)
table(Novershtern.fine$labels) # Number cells with same annotation; grouped
table(Novershtern.fine$pruned.labels) # Low-quality annotations removed
Harmonized_CNS_1[["Novershtern.fine"]] <- Novershtern.fine$labels
Harmonized_CNS_1 <- SetIdent(Harmonized_CNS_1, value = "Novershtern.fine")
UMAPPlot(Harmonized_CNS_1, label = TRUE, repel = TRUE) + NoLegend()
```


# 13. Preliminary differential gene expression

Expected gene expression? Aberrant gene expression? Can we identify the myeloid cluster(s) based on DEGs?

- *FindAllMarkers*. Default logfc.threshold = 0.25. Default test.use = "wilcox". Default min.pct (only test genes detected in a minimum fraction of cells) = 0.1.

```{r, fig.dim = c(10, 7)}

Harmonized_CNS_1 <- SetIdent(Harmonized_CNS_1, value = Harmonized_CNS_1$seurat_clusters)
All_Markers <- FindAllMarkers(Harmonized_CNS_1, min.pct = 0.25, verbose = FALSE, only.pos = TRUE)
```

Top 10 cluster markers; sorted by average log2 fold change:
```{r, fig.dim = c(10, 7)}
options(max.print = 2000) # To overrule default max.print
as.data.frame(All_Markers %>% group_by(cluster) %>% slice_max(avg_log2FC, n = 10))
```


# 14. Finalization

```{r, fig.dim = c(10, 7)}

summary(Harmonized_CNS_1@meta.data)

for (i in CNS_1_Names) {
  Subset <- subset(Harmonized_CNS_1, orig.ident == i)
  print(i)
  summary(Subset@meta.data) %>%
    print()
}

UMAPPlot(Harmonized_CNS_1, label = TRUE)

table1 <- table(Harmonized_CNS_1@active.ident)
table1

str(Harmonized_CNS_1)
saveRDS(Harmonized_CNS_1, "CNS_1.rds")

sessionInfo()
```
